src/package-request.js

```diff
@@ -281,7 +281,8 @@ export default class PackageRequest {
       promises.push(this.resolver.find({
         pattern: depPattern,
         registry: remote.registry,
-        optional: false,
+        // dependencies of optional dependencies should themselves be optional
+        optional: this.optional,
```

Comment:
Pretty sure that would work fine because of [this code](https://github.com/yarnpkg/yarn/blob/6e7d396451ed7622a92f4a25d9370e5724c54635/src/package-reference.js#L112) but I'll add a unit test to prove it.